src/opencl_kde_depth_packet_processor.cl

```diff
@@ -0,0 +1,769 @@
+}
+
+
+void kernel processPixelStage2_phase3(global const float3 *a_in, global const float3 *b_in, global float *phase_1, global float *phase_2, global float *phase_3, global float *conf1, global float *conf2, global float *conf3)
+{
+  const uint i = get_global_id(0);
+
+  //read complex number real (a) and imaginary part (b)
+  float3 a = a_in[i];
+  float3 b = b_in[i];
+  
+  //calculate complex argument
+  float3 phase = atan2(b, a);
+  phase = select(phase, (float3)(0.0f), isnan(phase));
+  phase = select(phase, phase + 2.0f * M_PI_F, isless(phase, (float3)(0.0f)));
+
+  //calculate amplitude or the absolute value
+  float3 ir = sqrt(a * a + b * b) * AB_MULTIPLIER;
+  ir = select(ir, (float3)(0.0f), isnan(ir));
+  
+  float ir_sum = ir.x + ir.y + ir.z;
+
+  float phase_first = 0.0;
+  float phase_second = 0.0;
+  float phase_third = 0.0;
+  float J_1, J_2, J_3, unwrapping_likelihood1, unwrapping_likelihood2, unwrapping_likelihood3;
+
+  //scale with least common multiples of modulation frequencies
+  float3 t = phase / (2.0f * M_PI_F) * (float3)(3.0f, 15.0f, 2.0f);
+
+  float t0 = t.x;
+  float t1 = t.y;
+  float t2 = t.z;
+
+  //rank and extract three most likely phase hypothises
+  phaseUnWrapper3(t0, t1, t2, &phase_first, &phase_second, &phase_third, &J_1, &J_2, &J_3);
+
+  phase_1[i] = phase_first;
+  phase_2[i] = phase_second;
+  phase_3[i] = phase_third;
+
+  float phase_likelihood;
+  //check if near saturation
+  if(ir_sum < 0.4f * 65535.0f)
+  {
+    //calculate phase likelihood from amplitude
+    float var0, var1, var2;
+    calculatePhaseUnwrappingVar(ir, &var0, &var1, &var2);
+    phase_likelihood = exp(-(var0+var1+var2)/(2.0f * PHASE_CONFIDENCE_SCALE));
+    phase_likelihood = select(phase_likelihood, 0.0f, isnan(phase_likelihood));
+  }
+  else
+  {
+    phase_likelihood = 0.0f;
+  }
+
+  //merge unwrapping likelihood with phase likelihood
+  unwrapping_likelihood1 = phase_likelihood * exp(-J_1 / (2 * UNWRAPPING_LIKELIHOOD_SCALE));
+  unwrapping_likelihood2 = phase_likelihood * exp(-J_2 / (2 * UNWRAPPING_LIKELIHOOD_SCALE));
+  unwrapping_likelihood3 = phase_likelihood * exp(-J_3 / (2 * UNWRAPPING_LIKELIHOOD_SCALE));
+
+  //suppress confidence if phase is beyond allowed range
+  unwrapping_likelihood1 = phase_first > MAX_DEPTH * 9.0f / 18750.0f ? 0.0f: unwrapping_likelihood1;
+  unwrapping_likelihood2 = phase_second > MAX_DEPTH * 9.0f / 18750.0f ? 0.0f: unwrapping_likelihood2;
+  unwrapping_likelihood3 = phase_third > MAX_DEPTH * 9.0f / 18750.0f ? 0.0f: unwrapping_likelihood3;
+
+  conf1[i] = unwrapping_likelihood1;
+  conf2[i] = unwrapping_likelihood2;
+  conf3[i] = unwrapping_likelihood3;
+
+}
+
+
+
+void kernel filter_kde3(global const float *phase_1, global const float *phase_2, global const float *phase_3, global const float* conf1, global const float* conf2, global const float* conf3, global const float* gauss_filt_array, global const float* z_table, global const float* x_table, global float* depth)
```

Comment:
We need to synchronize after the phase hypothesis is extracted in processPixelStage2_phase.  In filter_kde we select the phase hypothesis given the KDE of  the neighbouring pixels and use the KDE-value for removing potential outliers. Perhaps the name "filter_kde" is misleading since this is more than a filter.